[MINOR][DOCS] Remove trailing slash from Conda cheat sheet URL in 'Using Conda' section to fix 404 error

onestn opened the pull request:
In the 'Using Conda' section of the Spark documentation, the cheat sheet URL includes a trailing slash, causing a 404 error when accessed. 

This commit removes the trailing slash to ensure the URL correctly directs to the intended page.

### What changes were proposed in this pull request?
The proposed change removes the trailing slash from the Conda cheat sheet URL in the 'Using Conda' section of the Spark documentation. This modification is intended to resolve the 404 error that occurs when the current URL with the trailing slash is accessed.


### Why are the changes needed?
The change is necessary because the existing URL with a trailing slash leads to a 404 error, hindering users' access to the Conda cheat sheet. Removing the slash will ensure the link directs users to the correct webpage, improving the documentation's usability.


### Does this PR introduce _any_ user-facing change?
No. This change only affects the documentation and does not alter the functionality of the Spark software itself.


### How was this patch tested?
This documentation fix was visually inspected to ensure the corrected URL resolves to the intended webpage without a 404 error. No additional code testing is necessary as this change is strictly within the documentation.


### Was this patch authored or co-authored using generative AI tooling?
No.

onestn committed:
Remove trailing slash from Conda cheat sheet URL in 'Using Conda' section to fix 404 error

In the 'Using Conda' section of the Spark documentation, the cheat sheet URL includes a trailing slash, causing a 404 error when accessed. This commit removes the trailing slash to ensure the URL correctly directs to the intended page.

LuciferYang committed:
[SPARK-45871][CONNECT] Optimizations collection conversion related to `.toBuffer` in the `connect` modules

### What changes were proposed in this pull request?
This PR includes the following optimizations related to `.toBuffer` in the `connect`s module:

1. For the two functions `sql(String, java.util.Map[String, Any]): DataFrame` and `sql(String, Array[_]): DataFrame` in `SparkSession`, the approach of using `.find` directly on the `CloseableIterator` to locate the target and utilizing `.foreach` to consume the remaining elements replaced the previous method of converting to a collection using `toBuffer.toSeq` and then searching for the target using `.find`. This approach avoids the need for an unnecessary collection creation.
2. For function `execute(proto.Relation.Builder => Unit): Unit` in `SparkSession`, as no elements are returned, `.foreach` is used instead of `.toBuffer` to avoid an unnecessary collection creation.
3. For function `execute(command: proto.Command): Seq[ExecutePlanResponse]` in `SparkSession`, in Scala 2.12, `s.c.TraversableOnce#toSeq` returns an `immutable.Stream`, which is a tail-lazy structure that may not consume all elements. Therefore, it is necessary to use `s.c.TraversableOnce#toBuffer` for materialization. However, in Scala 2.13, `s.c.IterableOnceOps#toSeq` constructs an `immutable.Seq`, which is not a lazy data structure and ensures consumption of all elements. Therefore, in Scala 2.13, `.toSeq` can be used directly to replace `toBuffer.toSeq`, saving an extra collection transformation.
4. The optimizations for the two functions `listAbandonedExecutions: Seq[ExecuteInfo]` and `listExecuteHolders` in `SparkConnectExecutionManager` are consistent with item 3 above.

Additionally, to prevent helper function used for testing from creating copies, a `private[connect]` scope helper function was added to the companion object of `ExecutePlanResponseReattachableIterator`.

### Why are the changes needed?
Avoid unnecessary collection copies

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Action
- Added new test cases to demonstrate that both `.toSeq` and `.foreach` can consume all elements in `ResponseReattachableIterator` in Scala 2.13.

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43745 from LuciferYang/SPARK-45871.

Authored-by: yangjie01 <yangjie01@baidu.com>
Signed-off-by: yangjie01 <yangjie01@baidu.com>

sahnib committed:
[SPARK-45655][SQL][SS] Allow non-deterministic expressions inside AggregateFunctions in CollectMetrics

### What changes were proposed in this pull request?

This PR allows non-deterministic expressions wrapped inside an `AggregateFunction` such as `count` inside `CollectMetrics` node. `CollectMetrics` is used to collect arbitrary metrics from the query, in certain scenarios user would like to collect metrics for filtering based on non-deterministic expressions (see query example below).

Currently, Analyzer does not allow non-deterministic expressions inside a `AggregateFunction` for `CollectMetrics`. This constraint is relaxed to allow collection of such metrics. Note that the metrics are relevant for a completed batch, and can change if the batch is replayed (because non-deterministic expression can behave differently for different runs).

While working on this feature, I found a issue with `checkMetric` logic to validate non-deterministic expressions inside an AggregateExpression. An expression is determined as non-deterministic if any of its children is non-deterministic, hence we need to match the case for `!e.deterministic && !seenAggregate` after we have matched if the current expression is a AggregateExpression. If the current expression is a AggregateExpression, we should validate further down in the tree recursively - otherwise we will fail for any non-deterministic expression.

```

val inputData = MemoryStream[Timestamp]

inputData.toDF()
      .filter("value < current_date()")
      .observe("metrics", count(expr("value >= current_date()")).alias("dropped"))
      .writeStream
      .queryName("ts_metrics_test")
      .format("memory")
      .outputMode("append")
      .start()

```

### Why are the changes needed?

1. Added a testcase to calculate dropped rows (by `CurrentBatchTimestamp`) and ensure the query is successful.

As an example, the query below fails (without this change) due to observe call on the DataFrame.

```

val inputData = MemoryStream[Timestamp]

inputData.toDF()
      .filter("value < current_date()")
      .observe("metrics", count(expr("value >= current_date()")).alias("dropped"))
      .writeStream
      .queryName("ts_metrics_test")
      .format("memory")
      .outputMode("append")
      .start()

```
2. Added testing in AnalysisSuite for non-deterministic expressions inside a AggregateFunction.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test cases added.

```

[warn] 20 warnings found
WARNING: Using incubator modules: jdk.incubator.vector, jdk.incubator.foreign
[info] StreamingQueryStatusAndProgressSuite:
09:14:39.684 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] No tests to run for hive / Test / testOnly
[info] - StreamingQueryProgress - prettyJson (436 milliseconds)
[info] - StreamingQueryProgress - json (3 milliseconds)
[info] - StreamingQueryProgress - toString (5 milliseconds)
[info] - StreamingQueryProgress - jsonString and fromJson (163 milliseconds)
[info] - StreamingQueryStatus - prettyJson (1 millisecond)
[info] - StreamingQueryStatus - json (1 millisecond)
[info] - StreamingQueryStatus - toString (2 milliseconds)
09:14:41.674 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /Users/bhuwan.sahni/workspace/spark/target/tmp/temporary-34d2749f-f4d0
-46d8-bc51-29da6411e1c5. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
09:14:41.710 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - progress classes should be Serializable (5 seconds, 552 milliseconds)
09:14:46.345 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /Users/bhuwan.sahni/workspace/spark/target/tmp/temporary-3a41d397-c3c1
-490b-9cc7-d775b0c42208. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
09:14:46.345 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - SPARK-19378: Continue reporting stateOp metrics even if there is no active trigger (1 second, 337 milliseconds)
09:14:47.677 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - SPARK-29973: Make `processedRowsPerSecond` calculated more accurately and meaningfully (455 milliseconds)
09:14:48.174 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /Users/bhuwan.sahni/workspace/spark/target/tmp/temporary-360fc3b9-a2c5
-430c-a892-c9869f1f8339. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
09:14:48.174 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - SPARK-45655: Use current batch timestamp in observe API (587 milliseconds)
09:14:48.768 WARN org.apache.spark.sql.streaming.StreamingQueryStatusAndProgressSuite:

```

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #43517 from sahnib/SPARK-45655.

Authored-by: Bhuwan Sahni <bhuwan.sahni@databricks.com>
Signed-off-by: Jungtaek Lim <kabhwan.opensource@gmail.com>

panbingkun committed:
[SPARK-45870][INFRA] Upgrade action/checkout to v4

### What changes were proposed in this pull request?
The pr aims to upgrade action/checkout from v3(node16) to v4(node20) (point to v4.1.1 now).

### Why are the changes needed?
- After V4.0.0, the action/checkout default runtime is node20.
   <img width="577" alt="image" src="https://github.com/apache/spark/assets/15246973/5e261225-9f72-46f8-b10a-7820919810b2">

- Node 16 has reached its [end of life](https://github.com/nodejs/Release/#end-of-life-releases)
   <img width="946" alt="image" src="https://github.com/apache/spark/assets/15246973/64a6c6bf-7b70-418e-b053-2b156484a200">
   https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #43744 from panbingkun/github_checkout_action.

Authored-by: panbingkun <pbk1982@gmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

LuciferYang committed:
[SPARK-45875][CORE] Remove `MissingStageTableRowData` from `core` module

### What changes were proposed in this pull request?
SPARK-15591(https://github.com/apache/spark/pull/13708) introduced the `MissingStageTableRowData`, but it is no longer used after SPARK-20648(https://github.com/apache/spark/pull/19698), so this PR removes it.

### Why are the changes needed?
Clean up unused code.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43748 from LuciferYang/SPARK-45875.

Authored-by: yangjie01 <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

MaxGekk committed:
[SPARK-45886][SQL] Output full stack trace in `callSite` of DataFrame context

### What changes were proposed in this pull request?
In the PR, I propose to include all available stack traces in DataFrame context to the `callSite` field and apparently to the `summary`. For now, DataFrame context contains only one item of stack trace, but later we'll add a config to control the number of items in stack traces (see https://github.com/apache/spark/pull/43695).

### Why are the changes needed?
To improve user experience with Spark SQL while debugging some issue. Users can see all available stack trace, and see from where the issue comes in user code from.

### Does this PR introduce _any_ user-facing change?
No, should not. Even if user's code parses the summary.

### How was this patch tested?
By running new test suite:
```
$ build/sbt "test:testOnly *QueryContextSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #43758 from MaxGekk/output-stack-trace.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

zhengruifeng committed:
[SPARK-45770][SQL][PYTHON][CONNECT] Introduce plan `DataFrameDropColumns` for `Dataframe.drop`

### What changes were proposed in this pull request?
Fix column resolution in DataFrame.drop

### Why are the changes needed?

```
from pyspark.sql.functions import col

# create first dataframe
left_df = spark.createDataFrame([(1, 'a'), (2, 'b'), (3, 'c')], ['join_key', 'value1'])
# create second dataframe
right_df = spark.createDataFrame([(1, 'aa'), (2, 'bb'), (4, 'dd')], ['join_key', 'value2'])
joined_df = left_df.join(right_df, on=left_df['join_key'] == right_df['join_key'], how='left')
display(joined_df)
cleaned_df = joined_df.drop(left_df['join_key'])
display(cleaned_df) # error here

JVM stacktrace:
org.apache.spark.sql.AnalysisException: [AMBIGUOUS_REFERENCE] Reference `join_key` is ambiguous, could be: [`join_key`, `join_key`]. SQLSTATE: 42704
	at org.apache.spark.sql.errors.QueryCompilationErrors$.ambiguousReferenceError(QueryCompilationErrors.scala:1957)
	at org.apache.spark.sql.catalyst.expressions.package$AttributeSeq.resolve(package.scala:377)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolve(LogicalPlan.scala:156)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveQuoted(LogicalPlan.scala:167)
	at org.apache.spark.sql.Dataset.$anonfun$drop$4(Dataset.scala:3071)
```

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
ci

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #43683 from zhengruifeng/sql_drop_plan.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

bersprockets committed:
[SPARK-45896][SQL] Construct `ValidateExternalType` with the correct expected type

### What changes were proposed in this pull request?

When creating a serializer for a `Map` or `Seq` with an element of type `Option`, pass an expected type of `Option`  to `ValidateExternalType` rather than the `Option`'s type argument.

### Why are the changes needed?

In 3.4.1, 3.5.0, and master, the following code gets an error:
```
scala> val df = Seq(Seq(Some(Seq(0)))).toDF("a")
val df = Seq(Seq(Some(Seq(0)))).toDF("a")
org.apache.spark.SparkRuntimeException: [EXPRESSION_ENCODING_FAILED] Failed to encode a value of the expressions: mapobjects(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -1), mapobjects(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -2), assertnotnull(validateexternaltype(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -2), IntegerType, IntegerType)), unwrapoption(ObjectType(interface scala.collection.immutable.Seq), validateexternaltype(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -1), ArrayType(IntegerType,false), ObjectType(class scala.Option))), None), input[0, scala.collection.immutable.Seq, true], None) AS value#0 to a row. SQLSTATE: 42846
...
Caused by: java.lang.RuntimeException: scala.Some is not a valid external type for schema of array<int>
  at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.MapObjects_0$(Unknown Source)
...

```
However, this code works in 3.3.3.

Similarly, this code gets an error:
```
scala> val df = Seq(Seq(Some(java.sql.Timestamp.valueOf("2023-01-01 00:00:00")))).toDF("a")
val df = Seq(Seq(Some(java.sql.Timestamp.valueOf("2023-01-01 00:00:00")))).toDF("a")
org.apache.spark.SparkRuntimeException: [EXPRESSION_ENCODING_FAILED] Failed to encode a value of the expressions: mapobjects(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -1), staticinvoke(class org.apache.spark.sql.catalyst.util.DateTimeUtils$, TimestampType, fromJavaTimestamp, unwrapoption(ObjectType(class java.sql.Timestamp), validateexternaltype(lambdavariable(MapObject, ObjectType(class java.lang.Object), true, -1), TimestampType, ObjectType(class scala.Option))), true, false, true), input[0, scala.collection.immutable.Seq, true], None) AS value#0 to a row. SQLSTATE: 42846
...
Caused by: java.lang.RuntimeException: scala.Some is not a valid external type for schema of timestamp
...
```
As with the first example, this code works in 3.3.3.

`SerializerBuildHelper#validateAndSerializeElement` will construct `ValidateExternalType` with an expected type of the `Option`'s type parameter. Therefore, for element types `Option[Seq/Date/Timestamp/BigDecimal]`, `ValidateExternalType` will try to validate that the element is of the contained type (e.g., `BigDecimal`) rather than of type `Option`. Since the element type is of type `Option`, the validation fails.

Validation currently works by accident for element types `Option[Map/<primitive-type]`, simply because in that case `ValidateExternalType` ignores that passed expected type and tries to validate based on the encoder's `clsTag` field (which, for the `OptionEncoder`, will be class `Option`).

### Does this PR introduce _any_ user-facing change?

Other than fixing the bug, no.

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #43770 from bersprockets/encoding_error.

Authored-by: Bruce Robbins <bersprockets@gmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

LuciferYang committed:
[MINOR][CONNECT][TESTS]  Use `Some(())` instead of `Some()` for `Option[Unit]` return type

### What changes were proposed in this pull request?
This pr fix a compile warning like

```
[warn] /Users/yangjie01/SourceCode/git/spark-mine-sbt/connector/connect/server/src/test/scala/org/apache/spark/sql/connect/plugin/SparkConnectPluginRegistrySuite.scala:95:9: adaptation of an empty argument list by inserting () is deprecated: this is unlikely to be what you want
[warn]         signature: Some.apply[A](value: A): Some[A]
[warn]   given arguments: <none>
[warn]  after adaptation: Some((): Unit)
[warn] Applicable -Wconf / nowarn filters for this fatal warning: msg=<part of the message>, cat=deprecation, site=org.apache.spark.sql.connect.plugin.ExampleCommandPlugin.process, origin=scala.Some.apply, version=2.11.0
[warn]     Some()
[warn]         ^
```

### Why are the changes needed?
Clean up a deprecated usage since Scala 2.11.0

### Does this PR introduce _any_ user-facing change?
No, just for test

### How was this patch tested?
Pass GitHub Action

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43733 from LuciferYang/minor-option-unit.

Lead-authored-by: yangjie01 <yangjie01@baidu.com>
Co-authored-by: YangJie <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

LuciferYang committed:
[SPARK-45874][SQL] Remove Java version check from `IsolatedClientLoader`

### What changes were proposed in this pull request?
This pr remove unnecessary Java version check from `IsolatedClientLoader`.

### Why are the changes needed?
Apache Spark 4.0.0 has a minimum requirement of Java 17, so the version check for Java 9 is not necessary.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43747 from LuciferYang/SPARK-45874.

Authored-by: yangjie01 <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

LuciferYang committed:
[SPARK-45830][CORE] Refactor `StorageUtils#bufferCleaner`

### What changes were proposed in this pull request?
This pr refactor `StorageUtils#bufferCleaner` as follows:
- Change the return value of `bufferCleaner` from `DirectBuffer => Unit` to `ByteBuffer => Unit`
- Directly calling `unsafe.invokeCleaner` instead of reflecting calls

### Why are the changes needed?
1. After Scala 2.13.9, it is recommended to use the `-release` instead of the `-target` for compilation. However, due to `sun.nio.ch` module was not exported, this can lead to the issue of class invisibility during Java cross compilation, such as building or testing using Java 21 with `-release:17` After this pr, the following compilation errors will not occur again when build core module using Java 21 with `-release:17`:

```
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala:71: object security is not a member of package sun
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:26: object nio is not a member of package sun
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:200: not found: type DirectBuffer
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:206: not found: type DirectBuffer
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:220: not found: type DirectBuffer
[ERROR] [Error] /Users/yangjie01/SourceCode/git/spark-mine-13/core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:26: Unused import
```

2. Direct use of `unsafe.invokeCleaner` provides better performance, compared to reflection calls, it is at least 30% faster

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Actions
- Manual check building core module using Java 21 with `-release:17`, no longer compilation failure logs above

Note: There is still an issue with other classes being invisible, which needs to be fixed in follow up

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43675 from LuciferYang/bufferCleaner.

Lead-authored-by: yangjie01 <yangjie01@baidu.com>
Co-authored-by: YangJie <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

LuciferYang committed:
[SPARK-45848][BUILD] Make `spark-version-info.properties` generated by `spark-build-info.ps1` include `docroot`

### What changes were proposed in this pull request?
The `spark-version-info.properties` generated by `spark-build-info` include `docroot=https://spark.apache.org/docs/latest`, this pr make `spark-version-info.properties` generated by `spark-build-info.ps1` also include `docroot` part.

### Why are the changes needed?
Keep the items generated by `spark-build-info` and `spark-build-info.ps1` consistent

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual verification with this pr:
- `spark-version-info.properties` generated by `spark-build-info`
```
version=4.0.0-SNAPSHOT
user=yangjie01
revision=d92f634bdaf8d040b7c7a5ca675db3cb265486b5
branch=SPARK-45848
date=2023-11-09T06:21:05Z
url=gitgithub.com:LuciferYang/spark.git
docroot=https://spark.apache.org/docs/latest
```
- `spark-version-info.properties` generated by `spark-build-info.ps1`
```
version=4.0.0-SNAPSHOT
user=yangjie01
revision=d92f634bdaf8d040b7c7a5ca675db3cb265486b5
branch=SPARK-45848
date=2023-11-09T06:22:25Z
url=gitgithub.com:LuciferYang/spark.git,
docroot=https://spark.apache.org/docs/latest
```

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43726 from LuciferYang/SPARK-45848.

Lead-authored-by: yangjie01 <yangjie01@baidu.com>
Co-authored-by: YangJie <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

MaxGekk committed:
[SPARK-45857][SQL] Enforce the error classes in sub-classes of `AnalysisException`

### What changes were proposed in this pull request?
In the PR, I propose to enforce creation of `AnalysisException` sub-class exceptions with an error class always. In particular, it converts the constructor with a message to private one, so, callers have to create a sub-class of `AnalysisException` with an error class.

### Why are the changes needed?
This simplifies migration on error classes.

### Does this PR introduce _any_ user-facing change?
No, since user code doesn't throw `AnalysisException` and its sub-classes in regular cases.

### How was this patch tested?
By existing test suites, for instance:
```
$ build/sbt "sql/testOnly *QueryParsingErrorsSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #43731 from MaxGekk/ban-message-subclasses-AnalysisException.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

heyihong committed:
[SPARK-45899][CONNECT] Set errorClass in errorInfoToThrowable

### What changes were proposed in this pull request?

- Set errorClass in errorInfoToThrowable

### Why are the changes needed?

- errorClass should be set even when error enrichment is not working

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

`build/sbt "connect-client-jvm/testOnly *ClientE2ETestSuite"`

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #43772 from heyihong/SPARK-45899.

Authored-by: Yihong He <heyihong.cn@gmail.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

cdkrot committed:
[SPARK-45733][CONNECT][PYTHON] Support multiple retry policies

### What changes were proposed in this pull request?

Support multiple retry policies defined at the same time. Each policy determines which error types it can retry and how exactly those should be spread out.

### Why are the changes needed?

Different error types should be treated differently For instance, networking connectivity errors and remote resources being initialized should be treated separately.

### Does this PR introduce _any_ user-facing change?
No (as long as user doesn't poke within client internals).

### How was this patch tested?
Unit tests, some hand testing.

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43591 from cdkrot/SPARK-45733.

Authored-by: Alice Sayutina <alice.sayutina@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

HyukjinKwon commented:
Mind taking a quick look at https://github.com/apache/spark/pull/43763/checks?check_run_id=18583834991? LGTM but would like to make sure the CI passes.

dongjoon-hyun committed:
[SPARK-45901][DOCS] Collect and update `spark-standalone.md` with new confs

### What changes were proposed in this pull request?

This PR aims to collect and make `spark-standalone.md` up-to-date with new configurations to improve UX.

### Why are the changes needed?

By collecting and syncing with the code, the users can find all informations in a single page.

1. For example, the existing `zookeeper`-related configuration in `Configuration -> Deploy` section although it's only applicable in `Spark Standalone` cluster. In addition, it's partially described. We had better move it to `Spark Standalone` page.
    - From: https://spark.apache.org/docs/latest/configuration.html#deploy
    - To: https://spark.apache.org/docs/latest/spark-standalone.html

2. Add missing configurations.
    - spark.dead.worker.persistence

3. Revise the default value
   - spark.deploy.defaultCores

3. Add all Spark 4.0.0 configurations.
    - spark.deploy.maxDrivers
    - spark.deploy.appNumberModulo
    - spark.deploy.driverIdPattern
    - spark.deploy.appIdPattern
    - spark.worker.idPattern

### Does this PR introduce _any_ user-facing change?

Yes, but this is a documentation only.

### How was this patch tested?

Manual review.

**CHANGED PART 1**
<img width="925" alt="Screenshot 2023-11-12 at 4 18 21 PM" src="https://github.com/apache/spark/assets/9700541/266f88dc-0fe2-4378-82a9-502fd8374737">

**CHANGED PART 2**
<img width="1444" alt="Screenshot 2023-11-12 at 4 18 08 PM" src="https://github.com/apache/spark/assets/9700541/c19f976e-f3b3-43b1-874f-46618fa19997">

**CHANGED PART 3**
<img width="924" alt="Screenshot 2023-11-12 at 4 54 38 PM" src="https://github.com/apache/spark/assets/9700541/5e6d620f-5a1e-44d9-92f8-b7c0f61e27a3">

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #43774 from dongjoon-hyun/SPARK-45901.

Authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

HyukjinKwon committed:
[MINOR][PYTHON] Better error message when Python worker crushes

### What changes were proposed in this pull request?

This PR improves the Python UDF error messages to be more actionable.

### Why are the changes needed?

Suppose you face a segfault error:

```python
from pyspark.sql.functions import udf
import ctypes
spark.range(1).select(udf(lambda x: ctypes.string_at(0))("id")).collect()
```

The current error message is not actionable:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
get_return_value
    raise Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling o82.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 15 in stage 1.0 failed 1 times, most recent failure: Lost task 15.0 in stage 1.0 (TID 31) (192.168.123.102 executor driver): org.apache.spark.SparkException:
Python worker exited unexpectedly (crashed)
```

After this PR, it fixes the error message as below:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
get_return_value
    raise Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling o59.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 15 in stage 0.0 failed 1 times, most recent failure: Lost task 15.0 in stage 0.0 (TID 15) (192.168.123.102 executor driver): org.apache.spark.SparkException:
Python worker exited unexpectedly (crashed). Consider setting 'spark.sql.execution.pyspark.udf.faulthandler.enabled'
or 'spark.python.worker.faulthandler.enabled' configuration to 'true' forthe better Python traceback.
```

So you can try this out

```python
from pyspark.sql.functions import udf
import ctypes
spark.conf.set("spark.sql.execution.pyspark.udf.faulthandler.enabled", "true")
spark.range(1).select(udf(lambda x: ctypes.string_at(0))("id")).collect()
```

that now shows where the segfault happens:

```
Caused by: org.apache.spark.SparkException: Python worker exited unexpectedly (crashed): Fatal Python error: Segmentation fault

Current thread 0x00007ff84ae4b700 (most recent call first):
  File "/.../envs/python3.9/lib/python3.9/ctypes/__init__.py", line 525 in string_at
  File "<stdin>", line 1 in <lambda>
  File "/.../lib/pyspark.zip/pyspark/util.py", line 88 in wrapper
  File "/.../lib/pyspark.zip/pyspark/worker.py", line 99 in <lambda>
  File "/.../lib/pyspark.zip/pyspark/worker.py", line 1403 in <genexpr>
  File "/.../lib/pyspark.zip/pyspark/worker.py", line 1403 in mapper
```

### Does this PR introduce _any_ user-facing change?

Yes, it fixes the error message actionable.

### How was this patch tested?

Manually tested as above.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #43778 from HyukjinKwon/minor-error-improvement.

Authored-by: Hyukjin Kwon <gurwls223@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

zhengruifeng committed:
[SPARK-45859][ML] Make UDF objects in ml.functions lazy

### What changes were proposed in this pull request?
Since JVM runs static codes only once, if loading functions$ fails, it will always report java.lang.NoClassDefFoundError: Could not initialize class
```
23/11/01 23:06:21 WARN TaskSetManager: Lost task 136.0 in stage 9565.0 (TID 4557384) (10.4.35.209 executor 16): TaskKilled (Stage cancelled: Job aborted due to stage failure: Task 2 in stage 9565.0 failed 4 times, most recent failure: Lost task 2.3 in stage 9565.0 (TID 4558369) (10.4.56.6 executor 71): java.io.IOException: unexpected exception type
	at java.io.ObjectStreamClass.throwMiscException(ObjectStreamClass.java:1750)
	at java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1280)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2222)
	…
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:900)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:795)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor520.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.lang.invoke.SerializedLambda.readResolve(SerializedLambda.java:230)
	at sun.reflect.GeneratedMethodAccessor224.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1274)
	... 388 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.apache.spark.ml.functions$
	... 396 more
```

This PR just changes `functions.*` as lazy avoid hitting this issue because the initialization codes of a lazy val is not in static codes.

### Why are the changes needed?
to fix a intermittent bug

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
new UT

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #43739 from zhengruifeng/ml_lazy_udfs.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

zhengruifeng committed:
[SPARK-45723][PYTHON][CONNECT][FOLLOWUP] Replace `toPandas` with `_to_table` in catalog methods

### What changes were proposed in this pull request?
followup of https://github.com/apache/spark/pull/43583, replace `toPandas` with `_to_table` in catalog methods

### Why are the changes needed?
pandas conversion not needed

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
ci

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #43780 from zhengruifeng/py_catalog_arrow_followup.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

pan3793 committed:
[SPARK-45906][YARN] Fix error message extraction from ResourceNotFoundException

### What changes were proposed in this pull request?

This PR aims to fix the error message extraction from `ResourceNotFoundException`, the current wrong implementation also has a potential NPE issue.

### Why are the changes needed?

This bug is introduced in SPARK-43202, previously, `e.getCause()` is used to unwrap `InvocationTargetException`, after replacing reflection invocation with direct API calling, we should not apply `getCause()`.

### Does this PR introduce _any_ user-facing change?

Yes, bug fix.

### How was this patch tested?

Review.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #43782 from pan3793/SPARK-45906.

Authored-by: Cheng Pan <chengpan@apache.org>
Signed-off-by: Kent Yao <yao@apache.org>

allisonwang-db committed:
[SPARK-45783][PYTHON][CONNECT] Improve error messages when Spark Connect mode is enabled but remote URL is not set

### What changes were proposed in this pull request?

This PR improves the error messages when `SPARK_CONNECT_MODE_ENABLED` is defined but neither `spark.remote` option nor the `SPARK_REMOTE` env var is set.

### Why are the changes needed?

To improve the error message. Currently the error looks like a bug:
```
url = opts.get("spark.remote", os.environ.get("SPARK_REMOTE"))

>          if url.startswith("local"):
E          AttributeError: 'NoneType' object has no attribute 'startswith'
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New unit test

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #43653 from allisonwang-db/spark-45783-fix-url-err.

Authored-by: allisonwang-db <allison.wang@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

Commit:
Trigger CI

Commit:
Trigger CI

LuciferYang committed:
[SPARK-45902][SQL] Remove unused function `resolvePartitionColumns` from `DataSource`

### What changes were proposed in this pull request?
`resolvePartitionColumns` was introduced by SPARK-37287 (https://github.com/apache/spark/pull/37099) and become unused after SPARK-41713 (https://github.com/apache/spark/pull/39220), so this pr remove it from `DataSource`.

### Why are the changes needed?
Clean up unused code.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions.

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #43779 from LuciferYang/SPARK-45902.

Lead-authored-by: yangjie01 <yangjie01@baidu.com>
Co-authored-by: YangJie <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

yaooqinn committed:
[SPARK-45907][CORE] Use  Java9+  ProcessHandle APIs to computeProcessTree in ProcfsMetricsGetter

### What changes were proposed in this pull request?

This PR uses  Java9+ ProcessHandle APIs to computeProcessTree in ProcfsMetricsGetter.

### Why are the changes needed?

Simplify the code

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests
### Was this patch authored or co-authored using generative AI tooling?

no

Closes #43783 from yaooqinn/SPARK-45907.

Authored-by: Kent Yao <yao@apache.org>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

wangyum committed:
[SPARK-45909][SQL] Remove `NumericType` cast if it can safely up-cast in `IsNotNull`

### What changes were proposed in this pull request?

Similar to SPARK-37922. We can remove the cast if it can safely up-cast in `IsNotNull`.

### Why are the changes needed?

Improve the query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #43785 from wangyum/SPARK-45909.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

xiliang678 committed:
[SPARK-45892][SQL] Refactor optimizer plan validation to decouple `validateSchemaOutput` and `validateExprIdUniqueness`

### What changes were proposed in this pull request?

Currently, the expressionIDUniquenessValidation is [closely coupled with outputSchemaValidation](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala#L403C7-L411C8).

This PR refactors the code to improve readability and maintainability.

### Why are the changes needed?

Improve code readability and maintainability.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #43761 from xil-db/SPARK-45892-validation-refactor.

Authored-by: Xi Liang <xi.liang@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

ulysses-you committed:
[SPARK-45882][SQL] BroadcastHashJoinExec propagate partitioning should respect CoalescedHashPartitioning

### What changes were proposed in this pull request?

Add HashPartitioningLike trait and make HashPartitioning and CoalescedHashPartitioning extend it. When we propagate output partiitoning, we should handle HashPartitioningLike instead of HashPartitioning. This pr also changes the BroadcastHashJoinExec to use HashPartitioningLike to avoid regression.

### Why are the changes needed?

Avoid unnecessary shuffle exchange.

### Does this PR introduce _any_ user-facing change?

yes, avoid regression

### How was this patch tested?

add test

### Was this patch authored or co-authored using generative AI tooling?

no

Closes #43753 from ulysses-you/partitioning.

Authored-by: ulysses-you <ulyssesyou18@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

WweiL committed:
[SPARK-45845][SS][UI] Add number of evicted state rows to streaming UI

### What changes were proposed in this pull request?

In a stateful query, a watermark has two responsibilities:
1. drop late rows
2. Evict state rows from state store

Before we only log "aggregated number of rows dropped by watermark". This is case 1. But people would confuse this with case 2. This PR purpose we also add a chart for case 2. Also made the explanation for case 1 more verbose.

Now:
<img width="1063" alt="image" src="https://github.com/apache/spark/assets/10248890/b1f7f94f-94fe-490b-b2d9-abbb7170b67b">

### Why are the changes needed?

UI improvement

### Does this PR introduce _any_ user-facing change?

Yes, users will be able to view the new UI

### How was this patch tested?

Manually tested

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #43723 from WweiL/streaming-ui-update.

Authored-by: Wei Liu <wei.liu@databricks.com>
Signed-off-by: Jungtaek Lim <kabhwan.opensource@gmail.com>

Commit:
Trigger CI

HyukjinKwon commented:
Oops, seems like sth went wrong. Feel free to close and create a new PR if that's easier.

dead-1ine commented:
I'll close this PR and reopen it. Thank you :)
